In [1]:
import pandas as pd

In [2]:
pip install fuzzywuzzy

Note: you may need to restart the kernel to use updated packages.


In [3]:
file_path = r'C:\Users\capri\Workspace\Python cheatsheet\wind_turbine\EIA_2017_2021_load.csv'

In [4]:
df = pd.read_csv(file_path, low_memory= False)

In [5]:
df.head()

Plant Id Combined Heat And\nPower Plant Nuclear Unit Id    Plant Name  \
0         2                              N               .  Bankhead Dam   
1         3                              N               .         Barry   
2         3                              N               .         Barry   
3         3                              N               .         Barry   
4         3                              N               .         Barry   

      Operator Name Operator Id Plant State Census Region NERC Region  \
0  Alabama Power Co         195          AL           ESC        SERC   
1  Alabama Power Co         195          AL           ESC        SERC   
2  Alabama Power Co         195          AL           ESC        SERC   
3  Alabama Power Co         195          AL           ESC        SERC   
4  Alabama Power Co         195          AL           ESC        SERC   

   NAICS Code  ...  Total Source Station_Use Direct_Use Total\n Facility Use  \
0          22  ...           NaN         NaN        NaN                  NaN   
1          22  ...           NaN         NaN        NaN                  NaN   
2          22  ...           NaN         NaN        NaN                  NaN   
3          22  ...           NaN         NaN        NaN                  NaN   
4          22  ...           NaN         NaN        NaN                  NaN   

  Retail Sales Sales\n for Resale Tolling\n Agreements Outgoing\n Electricity  \
0          NaN                NaN                  NaN                    NaN   
1          NaN                NaN                  NaN                    NaN   
2          NaN                NaN                  NaN                    NaN   
3          NaN                NaN                  NaN                    NaN   
4          NaN                NaN                  NaN                    NaN   

  Total_Disposition Revenue from Resale  
0               NaN                 NaN  
1               NaN                 NaN  
2               NaN                 NaN  
3               NaN                 NaN  
4               NaN                 NaN  

[5 rows x 34 columns]

In [6]:
pd.DataFrame(df.dtypes, columns = ['Datatype'])

Datatype
Plant Id                               int64
Combined Heat And\nPower Plant        object
Nuclear Unit Id                       object
Plant Name                            object
Operator Name                         object
Operator Id                           object
Plant State                           object
Census Region                         object
NERC Region                           object
NAICS Code                             int64
EIA Sector Number                      int64
Sector Name                           object
Reported\nPrime Mover                 object
Reported\nFuel Type Code              object
AER\nFuel Type Code                   object
Physical\nUnit Label                  object
Total Fuel Consumption\nQuantity      object
Electric Fuel Consumption\nQuantity   object
Total Fuel Consumption\nMMBtu         object
Elec Fuel Consumption\nMMBtu          object
Net Generation\n(Megawatthours)       object
YEAR                                  object
Gross\n Generation                    object
Incoming\n Electricity                object
Total Source                          object
Station_Use                           object
Direct_Use                            object
Total\n Facility Use                  object
Retail Sales                          object
Sales\n for Resale                    object
Tolling\n Agreements                  object
Outgoing\n Electricity                object
Total_Disposition                     object
Revenue from Resale                   object

In [7]:
df['Operator Name'] = df['Operator Name'].astype(str)

In [8]:
from fuzzywuzzy import process

company_list = ['Amazon Wind','Arbor Hill', 'Acciona','AES','Beebe','Black Hills','Blackstone','Blue Canyon Windpower',
                'Blue Summit','Brady Wind','Cassia','Cedar Springs Wind','CHI','Cimarron','Consolidated Edison',
                'Creston Ridge','DTE','Duke Energy','Echo','EDF Renewable','Enel','Eurus Combine Hills','Evergy',
                'Fluvanna Wind Energy','Foundation CA Fund','FPL Energy','Golden Hills','Green Pastures','Harvest',
                'Heartland','Hidalgo Wind Farm','Iowa Lakes','Lincoln Electric','Longroad Energy Services',
                'Los Vientos Windpower','Madison','Massachusetts','Meadow Lake Wind Farm','Mesquite','Michigan Wind',
                'Musselshell','NaturEner','NextEra','NorthWestern Energy','NRG','OEE','Osceola','Paulding Wind Farm','Pecos',
                'Pinyon Pines','Power County Wind Park','Rattlesnake','Rising Tree Wind Farm','Rock County']

def replace_similar_values(value):
    for standard_name in company_list:
        if process.extractOne(value, [standard_name])[1] > 80:
            return standard_name
    return value

C:\Users\capri\anaconda3\lib\site-packages\fuzzywuzzy\fuzz.py:11: UserWarning: Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning
  warnings.warn('Using slow pure-python SequenceMatcher. Install python-Levenshtein to remove this warning')


In [9]:
# function call
df['Operator Name'] = df['Operator Name'].apply(replace_similar_values)

In [10]:
#df.to_csv('EIA_2017_2021_clean.csv', index = False)

In [11]:
# load the wind turbine file

file_path2 = r'C:\Users\capri\Workspace\Python cheatsheet\wind_turbine\wind_turbine_20220114.csv'


In [12]:
df_wind = pd.read_csv(file_path2, encoding='latin-1')


In [13]:
df_wind.head()

case_id    faa_ors           faa_asn  usgs_pr_id   eia_id t_state  \
0  3072661        NaN               NaN      5149.0  52161.0      CA   
1  3072695        NaN               NaN      5143.0  52161.0      CA   
2  3072704        NaN               NaN      5146.0  52161.0      CA   
3  3063272  19-028134  2014-WTE-4084-OE         NaN      NaN      IA   
4  3053390  19-028015  2015-WTE-6386-OE         NaN      NaN      IA   

       t_county  t_fips               plant_name  p_year  ...     t_rsa  \
0   Kern County    6029                 251 Wind  1987.0  ...       NaN   
1   Kern County    6029                 251 Wind  1987.0  ...       NaN   
2   Kern County    6029                 251 Wind  1987.0  ...       NaN   
3  Story County   19169  30 MW Iowa DG Portfolio  2017.0  ...  12271.85   
4  Boone County   19015  30 MW Iowa DG Portfolio  2017.0  ...  12271.85   

   t_ttlh retrofit retrofit_year  t_conf_atr  t_conf_loc  t_img_date  \
0     NaN        0           NaN           2           3    5/8/2018   
1     NaN        0           NaN           2           3    5/8/2018   
2     NaN        0           NaN           2           3    5/8/2018   
3   150.0        0           NaN           3           3   4/24/2017   
4   150.0        0           NaN           3           3    6/1/2017   

      t_img_srce       xlong       ylat  
0  Digital Globe -118.363762  35.077908  
1  Digital Globe -118.364410  35.077435  
2  Digital Globe -118.364197  35.077644  
3  Digital Globe  -93.430367  42.028233  
4  Digital Globe  -93.700424  41.977608  

[5 rows x 27 columns]

In [14]:
# check EIA ID for nulls- there are 5793 null EIA ID values

missing_values = df_wind.isnull().sum().sort_values(ascending=False)
missing_values.head(10)

retrofit_year    64822
usgs_pr_id       32545
t_img_date        8316
t_ttlh            6180
t_height_m        6180
t_rsa             5934
t_diameter_m      5934
eia_id            5793
faa_ors           5792
t_model           5777
dtype: int64

In [15]:
df['Plant Name'] = df['Plant Name'].astype(str)

In [16]:
pd.DataFrame(df_wind.dtypes, columns = ['Datatype'])

Datatype
case_id          int64
faa_ors         object
faa_asn         object
usgs_pr_id     float64
eia_id         float64
t_state         object
t_county        object
t_fips           int64
plant_name      object
p_year         float64
p_tnum           int64
plant_cap_MW   float64
t_manu          object
t_model         object
t_cap          float64
t_height_m     float64
t_diameter_m   float64
t_rsa          float64
t_ttlh         float64
retrofit         int64
retrofit_year  float64
t_conf_atr       int64
t_conf_loc       int64
t_img_date      object
t_img_srce      object
xlong          float64
ylat           float64

In [17]:
df_wind['plant_name'] = df_wind['plant_name'].astype(str)

In [18]:
df['Plant Id'] = df['Plant Id'].astype('float64')

#### We have the 2 dataframes now loaded in Python. we can fill the Null EIA_id with matching values from EIA 2017-2021 file by using Numpy.where function. We can also use pd.merge function via left and right join, but its best to keep the 2 dataframes separate to be loaded into Tableau. Later join can be created in Tableau using eia_id = Plant Id.

#### First we need to create a copy of df as df_eia and remove duplicate plant names in df_wind

In [19]:
df_eia = df.copy()

In [32]:
df_eia = df_eia.reset_index()

#### Lets filter df_wind by null eia_ids and check if there are duplicate plant name within the nulls

In [33]:
null_eia_id = df_wind[df_wind['eia_id'].isnull()]

duplicate_plant_names = null_eia_id[null_eia_id.duplicated(subset='plant_name', keep=False)]

In [34]:
print(null_eia_id['plant_name'].value_counts())

unknown Tehachapi Wind Resource Area 1    712
unknown Torrance County                   355
unknown Crockett County                   180
Patterson Pass                            132
TB Flats I & II                           132
                                         ... 
Lakeview Ridge                              1
Lake Region State College Wind Turbine      1
Lake Region Community Hybrid, LLC           1
unknown Canadian County                     1
Town of Richland                            1
Name: plant_name, Length: 458, dtype: int64


#### Now lets lookup eia_id from df_eia and use it to fill Null values in df_wind

In [30]:
import numpy as np

In [38]:
# Remove duplicates from the 'Plant Name' column in df
df_eia = df.drop_duplicates(subset='Plant Name', keep = False)


# Update values in df_wind.eia_id based on the condition
df_wind['eia_id'] = np.where(df_wind['eia_id'].isnull(),
    df_wind['plant_name'].map(df_eia.set_index('Plant Name')['Plant Id']),
    df_wind['eia_id']
)

In [39]:
df_wind.head()         

case_id    faa_ors           faa_asn  usgs_pr_id   eia_id t_state  \
0  3072661        NaN               NaN      5149.0  52161.0      CA   
1  3072695        NaN               NaN      5143.0  52161.0      CA   
2  3072704        NaN               NaN      5146.0  52161.0      CA   
3  3063272  19-028134  2014-WTE-4084-OE         NaN      NaN      IA   
4  3053390  19-028015  2015-WTE-6386-OE         NaN      NaN      IA   

       t_county  t_fips               plant_name  p_year  ...     t_rsa  \
0   Kern County    6029                 251 Wind  1987.0  ...       NaN   
1   Kern County    6029                 251 Wind  1987.0  ...       NaN   
2   Kern County    6029                 251 Wind  1987.0  ...       NaN   
3  Story County   19169  30 MW Iowa DG Portfolio  2017.0  ...  12271.85   
4  Boone County   19015  30 MW Iowa DG Portfolio  2017.0  ...  12271.85   

   t_ttlh retrofit retrofit_year  t_conf_atr  t_conf_loc  t_img_date  \
0     NaN        0           NaN           2           3    5/8/2018   
1     NaN        0           NaN           2           3    5/8/2018   
2     NaN        0           NaN           2           3    5/8/2018   
3   150.0        0           NaN           3           3   4/24/2017   
4   150.0        0           NaN           3           3    6/1/2017   

      t_img_srce       xlong       ylat  
0  Digital Globe -118.363762  35.077908  
1  Digital Globe -118.364410  35.077435  
2  Digital Globe -118.364197  35.077644  
3  Digital Globe  -93.430367  42.028233  
4  Digital Globe  -93.700424  41.977608  

[5 rows x 27 columns]

In [40]:
missing_values = df_wind.isnull().sum().sort_values(ascending=False)
missing_values.head(10)

retrofit_year    64822
usgs_pr_id       32545
t_img_date        8316
t_ttlh            6180
t_height_m        6180
t_rsa             5934
t_diameter_m      5934
faa_ors           5792
t_model           5777
eia_id            5688
dtype: int64

#### There are still 5688 null values in eia_id column. Since these are in a very small proportion of the total data, we will ignore these

In [41]:
df_wind.to_csv('USwind_turbine_2017_2021_clean.csv', index = False)

#### These 2 csv files are now ready to be loaded in Tableau